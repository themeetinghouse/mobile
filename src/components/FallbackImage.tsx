import React, { useState } from 'react';
import { Image, ImageStyle, ImageBackground, Animated } from 'react-native';

interface Props {
  uri: string;
  catchUri: string;
  style?: ImageStyle;
}

export default function FallbackImage({
  uri,
  catchUri,
  style,
}: Props): JSX.Element {
  const [source, setSource] = useState(uri);

  return (
    <Image
      testID="fallback-image"
      source={{ uri: source }}
      style={style}
      onError={() => setSource(catchUri)}
    />
  );
}

interface AnimatedProps extends Pick<Props, 'uri' | 'catchUri'> {
  style: Animated.WithAnimatedObject<ImageStyle>;
}

export function AnimatedFallbackImage({
  uri,
  catchUri,
  style,
}: AnimatedProps): JSX.Element {
  const [source, setSource] = useState(uri);

  return (
    <Animated.Image
      testID="fallback-image"
      source={{ uri: source }}
      style={style}
      onError={() => setSource(catchUri)}
    />
  );
}

interface PropsWithChildren extends Props {
  children: JSX.Element;
}

export function FallbackImageBackground({
  uri,
  catchUri,
  style,
  children,
}: PropsWithChildren): JSX.Element {
  const [source, setSource] = useState(uri);

  return (
    <ImageBackground
      testID="fallback-image"
      source={{ uri: source }}
      style={style}
      onError={() => setSource(catchUri)}
    >
      {children}
    </ImageBackground>
  );
}
