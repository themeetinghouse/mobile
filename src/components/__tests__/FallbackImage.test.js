import React from 'react';
import { render, act } from '@testing-library/react-native';
import { Text } from 'react-native';
import FallbackImage, {
  AnimatedFallbackImage,
  FallbackImageBackground,
} from '../FallbackImage';

const uri = 'not-a-real-url.png';
const catchUri = 'fallback-uri';

const components = [
  <FallbackImage key={0} uri={uri} catchUri={catchUri} />,
  <AnimatedFallbackImage key={1} uri={uri} catchUri={catchUri} />,
  <FallbackImageBackground
    key={2}
    uri={uri}
    catchUri={catchUri}
    style={{ width: 100, height: 100 }}
  />,
];

describe('FallbackImage components', () => {
  components.forEach((component) => {
    test('expect uri, then catchUri on error', () => {
      const { queryByTestId } = render(component);

      expect(queryByTestId('fallback-image').props.source.uri).toEqual(uri);
      act(() => {
        queryByTestId('fallback-image').props.onError();
      });
      expect(queryByTestId('fallback-image').props.source.uri).toEqual(
        catchUri
      );
    });
  });
});

describe('FallbackImageBackground component', () => {
  test('can accept children', () => {
    const { toJSON } = render(
      <FallbackImageBackground
        key={2}
        uri={uri}
        catchUri={catchUri}
        style={{ width: 100, height: 100 }}
      >
        <Text>Test</Text>
      </FallbackImageBackground>
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
