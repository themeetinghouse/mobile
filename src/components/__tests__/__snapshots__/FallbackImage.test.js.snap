// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FallbackImageBackground component can accept children 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "height": 100,
      "width": 100,
    }
  }
>
  <Image
    onError={[Function]}
    source={
      Object {
        "uri": "not-a-real-url.png",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": 100,
          "width": 100,
        },
        undefined,
      ]
    }
    testID="fallback-image"
  />
  <Text>
    Test
  </Text>
</View>
`;
